From 333572e09148dc7dea674e1d47c3a4b5fa5e7047 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?S=C3=B6ren=20Tempel?= <tempel@uni-bremen.de>
Date: Thu, 4 Feb 2021 20:08:26 +0100
Subject: [PATCH] uri_parser: check if uri is long enough to even contain a ://

Before attempting to access these characters. This fixes an
out-of-bounds read on the provided URI buffer.
---
 sys/uri_parser/uri_parser.c                         |  2 +-
 tests/unittests/tests-uri_parser/tests-uri_parser.c | 11 +++++++++++
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/sys/uri_parser/uri_parser.c b/sys/uri_parser/uri_parser.c
index b71d022a9308..294c6d92cbec 100644
--- a/sys/uri_parser/uri_parser.c
+++ b/sys/uri_parser/uri_parser.c
@@ -55,7 +55,7 @@ static char *_consume_scheme(uri_parser_result_t *result, char *uri,
     result->scheme_len = p - uri;
 
     /* check if authority part exists '://' */
-    if ((p[1] != '\0') && (p[2] != '\0') && (p[1] == '/') && (p[2] == '/')) {
+    if (((uri_end - p) > 2) && (p[1] == '/') && (p[2] == '/')) {
         *has_authority = true;
         /* skip '://' */
         return p + 3;
diff --git a/tests/unittests/tests-uri_parser/tests-uri_parser.c b/tests/unittests/tests-uri_parser/tests-uri_parser.c
index 37d8147845d1..73772aba10f6 100644
--- a/tests/unittests/tests-uri_parser/tests-uri_parser.c
+++ b/tests/unittests/tests-uri_parser/tests-uri_parser.c
@@ -401,6 +401,17 @@ static const validate_t validate_uris[] = {
         "./this:that",
         "",
         0),
+    VEC("pP://",
+        true,
+        "pP",
+        "",
+        "",
+        "",
+        "",
+        "",
+        "",
+        "",
+        0),
 };
 
 static char _failure_msg[VEC_MSG_LEN];
